flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,298 @@
+		this.credentials = credentials;
+		this.maxMessagesToAcknowledge = maxMessagesToAcknowledge;
+		this.acknowledgeOnCheckpointFactory = acknowledgeOnCheckpointFactory;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		createAndSetAcknowledgeOnCheckpoint();
+		this.subscriber = pubSubSubscriberFactory.getSubscriber(credentials);
+		this.isRunning = true;
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	public void acknowledge(List<String> acknowledgementIds) {
+		if (!isRunning) {
+			return;
+		}
+
+		subscriber.acknowledge(acknowledgementIds);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		while (isRunning) {
+			try {
+				blockIfMaxMessagesToAcknowledgeLimitReached();
+
+				processMessage(sourceContext, subscriber.pull());
+			} catch (InterruptedException | CancellationException e) {
+				isRunning = false;
+			}
+		}
+		subscriber.close();
+	}
+
+	void processMessage(SourceContext<OUT> sourceContext, List<ReceivedMessage> messages) throws Exception {
+		synchronized (sourceContext.getCheckpointLock()) {
+			for (ReceivedMessage message : messages) {
+				acknowledgeOnCheckpoint.addAcknowledgeId(message.getAckId());
+
+				PubsubMessage pubsubMessage = message.getMessage();
+
+				OUT deserializedMessage = deserializationSchema.deserialize(pubsubMessage);
+				if (deserializationSchema.isEndOfStream(deserializedMessage)) {
+					cancel();
+					return;
+				}
+
+				sourceContext.collect(deserializedMessage);
+			}
+
+		}
+	}
+
+	private void blockIfMaxMessagesToAcknowledgeLimitReached() throws Exception {
+		while (maxMessagesToAcknowledge != NO_MAX_MESSAGES_TO_ACKNOWLEDGE_LIMIT && getOutstandingMessagesToAck() > maxMessagesToAcknowledge) {
+			LOG.debug("Sleeping because there are {} messages waiting to be ack'ed but limit is {}", getOutstandingMessagesToAck(), maxMessagesToAcknowledge);
+			Thread.sleep(100);
+		}
+	}
+
+	private Integer getOutstandingMessagesToAck() {
+		return acknowledgeOnCheckpoint.numberOfOutstandingAcknowledgements();
+	}
+
+	@Override
+	public void cancel() {
+		isRunning = false;
+	}
+
+	@Override
+	public TypeInformation<OUT> getProducedType() {
+		return deserializationSchema.getProducedType();
+	}
+
+	public static <OUT> PubSubSourceBuilder<OUT> newBuilder(DeserializationSchema<OUT> deserializationSchema,
```

Comment:
These two public methods to get builders are not really a builder pattern. See the comments for the `PubSubSink`

Reply:
See the comment for PubSubSink